### ISSUE TITLE
Type error in `csma-pta.jani`.

### ISSUE THREAD

**Issue by koehlma** [ahartmanns/qcomp#103]
Thanks for assembling this great set of benchmarks! They just helped me a lot debugging my tool.

I believe that I found an error in one of the JANI files.

Line 96070 of [`pta/csma-pta/csma-pta.jani`](https://github.com/ahartmanns/qcomp/blob/master/benchmarks/pta/csma-pta/csma-pta.jani):
```json
"name": "M",
"type": "int",
"value": {
    "left": {
        "left": 2,
        "op": "pow",
        "right": "K"
    },
        "op": "-",
        "right": 1
    }
}
```
According to my understanding of the JANI-model specification the expression `pow(2, K) - 1` has type `real` because `pow(2, K)` has type `real`. However, a `real` cannot be assigned to an `int`.

**Comment by ahartmanns** [comment]
This is indeed a bug in the model according to the current jani-model spec. We can correct this manually in the JANI file, but since this was converted from a PRISM file by Storm, @tquatmann would you like to take a look if this is maybe a bug in the converter? Or would you like to make a case for changing the spec (which we should be _very_ careful with by now)?

**Comment by davexparker** [comment]
At a glance, it looks like the JANI file matches the original PRISM one. In the PRISM modelling language, `pow(x,y)` has type `int` if both `x` and `y` are also `int`s.

**Comment by ahartmanns** [comment]
@davexparker, what happens in PRISM if `y` in `pow(x,y)` is negative? Then it can have a real (including irrational) result. I think this was the reason why we specified the result to be real, and in particular not make a special case that has to refer to the sign of the expression (which is hard to determine if `y` is an arbitrary expression, not a literal or constant).

**Comment by davexparker** [comment]
For an integer `pow(x,y)`,  a negative `y` results in a run-time evaluation error. If either `x` or `y` is a double, so is the result. So you'd use e.g. `pow(x,-1.0)` for the inverse. From memory, this is because powers are useful in integer variable ranges, and we don't have an operator for explicit casting from `double` to `int` (well you could use `round`, but that feels bit ugly).

**Comment by ahartmanns** [comment]
Ah, and in JANI, a numeric literal has type `int` if "it is an integer", so `-1.0` would have type `int` in JANI but not in PRISM.

I think we should just handle this as the difference it is: `pow(x, y)` in PRISM should be converted to the JANI equivalent of `(int)pow(x,y)` if `x` and `y` are integers, else it can stay `pow(x, y)`. And in the other direction, from JANI to PRISM, `(int)pow(x,y)` can become `pow(x,y)` if `x` and `y` are `int` and provably `y >= 0`, else there probably should be an error (or generate PRISM with `round`/`ceil`/something like that).

**Comment by tquatmann** [comment]
I can confirm that this is an issue in the Storm's PRISM->JANI conversion. I will fix that in the Storm.

Just to be clear, in this example we would have something like: `{"op": "trc", "exp": { "left": 2, "op": "pow", "right": "K"}}`, correct?
In addition I will have a look through the existing PRISM models of QVBS to see whether this is also an issue in other models.

**Comment by ahartmanns** [comment]
Thanks!

> Just to be clear, in this example we would have something like: `{"op": "trc", "exp": { "left": 2, "op": "pow", "right": "K"}}`, correct?

I think so. Interesting that we decided to have only `floor` and `ceil` in basic JANI, with `trc` only in derived-operators. But I think most models use derived-operators anyway, so shouldn't be a problem.

**Comment by koehlma** [comment]
> In addition I will have a look through the existing PRISM models of QVBS to see whether this is also an issue in other models.

I found this issue by running [Momba](https://github.com/koehlma/momba/) (which performs type checking) on all QVBS models. I did not find this issue in any other model (that does not use model features besides `arrays`, `derived-operators`, and `trigonometric-functions`).

**Comment by davexparker** [comment]
@ahartmanns: Yes, adding an explicit cast in the PRISM->JANI direction makes sense.

For the JANI->PRISM direction, I guess you can just generally translate an integer cast in JANI to a PRISM `floor`. I don't see there is a need to explicitly reason about the case of `pow` or whether `y>=0` since the modeller would have had to explicitly write `(int)` for this situation to arise?

[Caveat: `(int)x` would actually become `x>0?floor(x):ceil(x)` I guess]

**Comment by tquatmann** [comment]
> I did not find this issue in any other model (that does not use model features besides arrays, derived-operators, and trigonometric-functions

Can you rephrase that @koehlma? For example, model feature `functions` is used sometimes. Did you check these too, or not?
Thank you very much for the effort!

**Comment by koehlma** [comment]
> Can you rephrase that @koehlma? For example, model feature `functions` is used sometimes. Did you check these too, or not?
> Thank you very much for the effort!

@tquatmann Sorry for the double negation madness.  No, I did not check these models. I only checked models that use `arrays`, `derived-operators`, and `trigonometric-functions` or any subset of these features. Momba does not support other model features like `functions` yet. However, I plan to add support for `functions`, `nondet-selection`, and `named-expressions` in the near future. Probably shortly after the CAV deadline.
